ocean_lib/common/ddo/credentials.py

```diff
@@ -0,0 +1,106 @@
+#
+# Copyright 2021 Ocean Protocol Foundation
+# SPDX-License-Identifier: Apache-2.0
+#
+from ocean_lib.common.agreements.consumable import ConsumableCodes, MalformedCredential
+
+
+class AddressCredential:
+    def __init__(self, asset):
+        self.asset = asset
+
+    def get_addresses_of_class(self, access_class="allow"):
+        """Get a filtered list of addresses from credentials (use with allow/deny)."""
+        entries = self.asset._credentials.get(access_class, [])
+
+        for entry in entries:
+            if entry["type"] == "address":
+                return [val.lower() for val in entry["values"]]
+
+        return []
+
+    def requires_credential(self):
+        allowed_addresses = self.get_addresses_of_class("allow")
+        denied_addresses = self.get_addresses_of_class("deny")
+
+        return allowed_addresses or denied_addresses
+
+    def validate_access(self, credential=None):
+        address = simplify_credential_to_address(credential)
+
+        allowed_addresses = self.get_addresses_of_class("allow")
+        denied_addresses = self.get_addresses_of_class("deny")
+
+        if not address and not self.requires_credential():
+            return ConsumableCodes.OK
+
+        if allowed_addresses and address.lower() not in allowed_addresses:
+            return ConsumableCodes.CREDENTIAL_NOT_IN_ALLOW_LIST
+
+        if not allowed_addresses and address.lower() in denied_addresses:
+            return ConsumableCodes.CREDENTIAL_IN_DENY_LIST
+
+        return ConsumableCodes.OK
+
+    def add_address_to_list_class(self, address, list_class="allow"):
+        address = address.lower()
+
+        if not self.asset._credentials or list_class not in self.asset._credentials:
+            self.asset._credentials[list_class] = [
+                {"type": "address", "values": [address]}
+            ]
+            return
+
+        address_entry = get_list_entry_with_type_address(
+            self.asset._credentials[list_class]
+        )
+
+        if not address_entry:
+            self.asset._credentials[list_class].append(
+                {"type": "address", "values": [address]}
+            )
+            return
+
+        address_entry = address_entry[0]
```

Comment:
Why is this line necessary?

Same question for the identical line in `remove_address_from_list_class`

Reply:
I didn't pluralise corectly. I need to test that there is an address type before I modify the one addres type. It will be clearer after my next commit.

Reply:
Ohh I see now.  `get_address_entry_of_class` currently returns a list from which you extract the 0th entry.

I think `get_address_entry_of_class` should extract the 0th entry before returning because there should only ever be 1 address entry in a given access class list.

```python
    def get_address_entry_of_class(self, access_class="allow"):
        """Get the address entry of a given credentials access class list. access_class = "allow" or "deny"."""
        entries = self.asset._credentials.get(access_class, [])
        address_entries = [entry for entry in entries if entry["type"] == "address"]
        return address_entries[0] if address_entries else None
```